src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SyntaxTokenToSemanticTokensEditHelper.cs

```diff
@@ -16,84 +16,109 @@ internal static class SyntaxTokenToSemanticTokensEditHelper
             SemanticTokens newTokens,
             IReadOnlyList<uint> previousResults)
         {
-            var oldData = previousResults;
-
-            if (oldData is null || oldData.Count == 0)
+            var diffs = UIntArrayDiffer.GetEdits(previousResults.ToArray(), newTokens.Data);
+            var edits = ComputeSemanticTokenEdits(diffs, newTokens.Data);
+            var result = new SemanticTokensEditCollection
             {
-                return newTokens;
-            }
+                ResultId = newTokens.ResultId,
+                Edits = edits,
+            };
 
-            var prevData = oldData;
-            var prevDataLength = oldData.Count;
-            var dataLength = newTokens.Data.Length;
-            var startIndex = 0;
-            while (startIndex < dataLength
-                && startIndex < prevDataLength
-                && prevData[startIndex] == newTokens.Data[startIndex])
-            {
-                startIndex++;
-            }
+            return result;
+        }
 
-            if (startIndex < dataLength && startIndex < prevDataLength)
+        private static IReadOnlyList<SemanticTokensEdit> ComputeSemanticTokenEdits(IReadOnlyList<DiffEdit> diffs, uint[] newArray)
+        {
+            var results = new List<SemanticTokensEdit>();
+            foreach (var diff in diffs)
```

Comment:
Spoke to @ryanbrandenburg offline but calling it out as a note to other reviewers. This is currently returning 1 edit per diff even though contiguous ones can be collated. The plan is to do that as a follow up.